forc-pkg/src/pkg.rs

```diff
@@ -1758,7 +1854,24 @@ fn dep_to_source(pkg_path: &Path, dep: &Dependency) -> Result<Source> {
                 let canonical_path = path.canonicalize().map_err(|e| {
                     anyhow!("Failed to canonicalize dependency path {:?}: {}", path, e)
                 })?;
-                Source::Path(canonical_path)
+                // Check if path is a member of a workspace.
+                let workspace_manifest = canonical_path
+                    .parent()
+                    .and_then(|parent_dir| WorkspaceManifestFile::from_dir(parent_dir).ok());
```

Comment:
Rather than doing all the I/O work of finding and reading this workspace again here (which will be attempted every time we visit a `Path` dependency) we should instead be checking our already loaded `MemberManifestFiles` set.

We can land as is for now, but we should open a high priority issue and make sure to address this soon after.

Reply:
created #3282 to track this